File: backend/routers/preview.py
"""Preview endpoints: full campaign structure with names and creative mapping."""
import json
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from database import get_db
from models.campaign import Project, AdSet, Ad, ProjectStatus
from models.creative import Creative
from services.naming import generate_all_names
from services.utm_builder import build_url_tags, preview_utm_url, get_utm_params_display
from services.creative_mapper import CreativeMapper, build_creatives_index

router = APIRouter(prefix="/api/preview", tags=["preview"])


@router.get("/{project_id}")
async def get_full_preview(project_id: int, db: AsyncSession = Depends(get_db)):
    """Get complete preview: campaign tree + generated names + creative mapping + UTMs."""
    result = await db.execute(
        select(Project)
        .where(Project.id == project_id)
        .options(selectinload(Project.ad_sets).selectinload(AdSet.ads))
    )
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")

    # Parse naming templates
    templates = json.loads(project.naming_template) if project.naming_template else None

    # Build ad_sets data for naming
    ad_sets_data = []
    for adset in project.ad_sets:
        ads_data = [
            {
                "name": ad.name,
                "creative_ref": ad.creative_ref or "",
                "format": "",
                "headline": ad.headline,
                "primary_text": ad.primary_text,
                "description": ad.description,
                "cta": ad.cta,
                "url": ad.url,
            }
            for ad in adset.ads
        ]
        ad_sets_data.append({
            "name": adset.name,
            "optimization_goal": adset.optimization_goal,
            "targeting_json": adset.targeting_json,
            "placements_json": adset.placements_json,
            "budget": adset.budget,
            "bid_strategy": adset.bid_strategy,
            "ads": ads_data,
        })

    # Generate names
    naming_result = generate_all_names(
        project_name=project.name,
        objective=project.campaign_objective or "OUTCOME_SALES",
        budget_type=project.budget_type.value if project.budget_type else "CBO",
        ad_sets=ad_sets_data,
        templates=templates,
    )

    # Update generated names in DB
    for i, adset in enumerate(project.ad_sets):
        if i < len(naming_result["ad_sets"]):
            adset.generated_name = naming_result["ad_sets"][i]["generated_name"]
            for j, ad in enumerate(adset.ads):
                if j < len(naming_result["ad_sets"][i]["ads"]):
                    ad.generated_name = naming_result["ad_sets"][i]["ads"][j]["generated_name"]
                    ad.url_tags = build_url_tags()

    await db.flush()

    # Build creative mapping
    creatives_result = await db.execute(
        select(Creative).where(Creative.project_id == project_id)
    )
    creatives = list(creatives_result.scalars().all())
    creative_list = [
        {"id": c.id, "base_name": c.base_name, "format": c.format.value if c.format else None,
         "original_name": c.original_name}
        for c in creatives
    ]
    index = build_creatives_index(creative_list)
    mapper = CreativeMapper(index)

    # Build preview
    url_tags = build_url_tags()
    base_url = project.destination_url or "https://example.com"

    preview = {
        "campaign": {
            "generated_name": naming_result["campaign_name"],
            "objective": project.campaign_objective,
            "budget_type": project.budget_type.value if project.budget_type else "CBO",
            "daily_budget": project.daily_budget,
            "lifetime_budget": project.lifetime_budget,
        },
        "utm_params": get_utm_params_display(url_tags),
        "utm_preview_url": preview_utm_url(base_url, url_tags),
        "ad_sets": [],
        "warnings": [],
    }

    for i, adset in enumerate(project.ad_sets):
        adset_preview = {
            "id": adset.id,
            "name": adset.name,
            "generated_name": adset.generated_name,
            "targeting": json.loads(adset.targeting_json) if adset.targeting_json else {},
            "optimization_goal": adset.optimization_goal,
            "budget": adset.budget,
            "ads": [],
        }

        for ad in adset.ads:
            mapping = mapper.get_mapping_for_ad(ad.creative_ref or "")
            ad_preview = {
                "id": ad.id,
                "name": ad.name,
                "generated_name": ad.generated_name,
                "creative_ref": ad.creative_ref,
                "headline": ad.headline,
                "primary_text": ad.primary_text,
                "cta": ad.cta,
                "url": ad.url,
                "url_tags": ad.url_tags,
                "creative_mapping": mapping["mappings"][:5],  # Top 5 placements
                "creative_warnings": mapping["warnings"],
            }
            adset_preview["ads"].append(ad_preview)
            preview["warnings"].extend(mapping["warnings"])

        preview["ad_sets"].append(adset_preview)

    project.status = ProjectStatus.previewed
    await db.commit()

    return preview


class NamingTemplateUpdate(BaseModel):
    campaign: Optional[str] = None
    adset: Optional[str] = None
    ad: Optional[str] = None


@router.put("/{project_id}/naming-template")
async def update_naming_template(
    project_id: int,
    templates: NamingTemplateUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update naming templates for a project."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")

    project.naming_template = json.dumps(templates.model_dump(exclude_unset=True))
    await db.commit()

    return {"status": "updated"}

File: api/lib/routes/preview.ts
import { Router, Request, Response } from "express";
import { eq } from "drizzle-orm";
import { db } from "../db/client.js";
import { projects, adSets, ads, creatives } from "../db/schema.js";

const router = Router();

// ── GET /:id ──────────────────────────────────────────────────────────────

router.get("/:id", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id, 10);

    // Load project
    const projectRows = await db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId))
      .limit(1);
    const project = projectRows[0];
    if (!project) {
      return res.status(404).json({ detail: "Project not found" });
    }

    // Load ad sets with ads
    const adSetRows = await db
      .select()
      .from(adSets)
      .where(eq(adSets.projectId, projectId));

    const adSetsWithAds = await Promise.all(
      adSetRows.map(async (adSet) => {
        const adRows = await db
          .select()
          .from(ads)
          .where(eq(ads.adSetId, adSet.id));
        return { ...adSet, ads: adRows };
      }),
    );

    // Parse naming templates
    const templates = project.namingTemplate
      ? JSON.parse(project.namingTemplate)
      : null;

    // Build ad_sets data for naming
    const adSetsData = adSetsWithAds.map((adSet) => ({
      name: adSet.name,
      optimization_goal: adSet.optimizationGoal,
      targeting_json: adSet.targetingJson,
      placements_json: adSet.placementsJson,
      budget: adSet.budget,
      bid_strategy: adSet.bidStrategy,
      ads: adSet.ads.map((ad) => ({
        name: ad.name,
        creative_ref: ad.creativeRef ?? "",
        format: "",
        headline: ad.headline,
        primary_text: ad.primaryText,
        description: ad.description,
        cta: ad.cta,
        url: ad.url,
      })),
    }));

    // Generate names using naming service
    let namingResult: any;
    try {
      const { generateAllNames } = await import("../services/naming.js");
      namingResult = generateAllNames(
        project.name,
        project.campaignObjective ?? "OUTCOME_SALES",
        project.budgetType ?? "CBO",
        adSetsData,
        templates,
      );
    } catch {
      // Fallback naming if service not available
      namingResult = {
        campaign_name: project.name,
        ad_sets: adSetsData.map((as_) => ({
          generated_name: as_.name,
          ads: as_.ads.map((ad) => ({
            generated_name: ad.name,
          })),
        })),
      };
    }

    // Update generated names in DB
    for (let i = 0; i < adSetsWithAds.length; i++) {
      const adSet = adSetsWithAds[i];
      const namingAdSet = namingResult.ad_sets?.[i];
      if (!namingAdSet) continue;

      await db
        .update(adSets)
        .set({ generatedName: namingAdSet.generated_name })
        .where(eq(adSets.id, adSet.id));
      adSet.generatedName = namingAdSet.generated_name;

      for (let j = 0; j < adSet.ads.length; j++) {
        const ad = adSet.ads[j];
        const namingAd = namingAdSet.ads?.[j];
        if (!namingAd) continue;

        let urlTags: string | null = null;
        try {
          const { buildUrlTags } = await import("../services/utm-builder.js");
          urlTags = buildUrlTags();
        } catch {
          // utm service not available
        }

        await db
          .update(ads)
          .set({
            generatedName: namingAd.generated_name,
            urlTags: urlTags,
          })
          .where(eq(ads.id, ad.id));
        ad.generatedName = namingAd.generated_name;
        ad.urlTags = urlTags;
      }
    }

    // Build creative mapping
    const creativeRows = await db
      .select()
      .from(creatives)
      .where(eq(creatives.projectId, projectId));

    const creativeList = creativeRows.map((c) => ({
      id: c.id,
      base_name: c.baseName,
      format: c.format,
      original_name: c.originalName,
    }));

    let mapper: any = null;
    try {
      const { buildCreativesIndex, CreativeMapper } = await import(
        "../services/creative-mapper.js"
      );
      const index = buildCreativesIndex(creativeList);
      mapper = new CreativeMapper(index);
    } catch {
      // service not available
    }

    // Build UTM display values
    let urlTags = "";
    let utmParamsDisplay: Record<string, string> = {};
    let utmPreviewUrl = "";
    const baseUrl = project.destinationUrl ?? "https://example.com";
    try {
      const utmModule = await import("../services/utm-builder.js");
      urlTags = utmModule.buildUrlTags();
      utmParamsDisplay = utmModule.getUtmParamsDisplay(urlTags);
      utmPreviewUrl = utmModule.previewUtmUrl(baseUrl, urlTags);
    } catch {
      // utm service not available
      utmPreviewUrl = baseUrl;
    }

    // Build preview response
    const warnings: string[] = [];

    const preview = {
      campaign: {
        generated_name: namingResult.campaign_name,
        objective: project.campaignObjective,
        budget_type: project.budgetType ?? "CBO",
        daily_budget: project.dailyBudget,
        lifetime_budget: project.lifetimeBudget,
      },
      utm_params: utmParamsDisplay,
      utm_preview_url: utmPreviewUrl,
      ad_sets: adSetsWithAds.map((adSet) => {
        const adsetPreview = {
          id: adSet.id,
          name: adSet.name,
          generated_name: adSet.generatedName,
          targeting: adSet.targetingJson ? JSON.parse(adSet.targetingJson) : {},
          optimization_goal: adSet.optimizationGoal,
          budget: adSet.budget,
          ads: adSet.ads.map((ad) => {
            let creativeMapping: any[] = [];
            let creativeWarnings: string[] = [];

            if (mapper) {
              const mapping = mapper.getMappingForAd(ad.creativeRef ?? "");
              creativeMapping = (mapping.mappings ?? []).slice(0, 5);
              creativeWarnings = mapping.warnings ?? [];
              warnings.push(...creativeWarnings);
            }

            return {
              id: ad.id,
              name: ad.name,
              generated_name: ad.generatedName,
              creative_ref: ad.creativeRef,
              headline: ad.headline,
              primary_text: ad.primaryText,
              cta: ad.cta,
              url: ad.url,
              url_tags: ad.urlTags,
              creative_mapping: creativeMapping,
              creative_warnings: creativeWarnings,
            };
          }),
        };
        return adsetPreview;
      }),
      warnings,
    };

    // Update project status
    await db
      .update(projects)
      .set({ status: "previewed" })
      .where(eq(projects.id, projectId));

    return res.json(preview);
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── PUT /:id/naming-template ──────────────────────────────────────────────

router.put("/:id/naming-template", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id, 10);

    const projectRows = await db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId))
      .limit(1);
    if (!projectRows[0]) {
      return res.status(404).json({ detail: "Project not found" });
    }

    // Accept { campaign, adset, ad } template strings
    const templateData: Record<string, string> = {};
    if (req.body.campaign !== undefined) templateData.campaign = req.body.campaign;
    if (req.body.adset !== undefined) templateData.adset = req.body.adset;
    if (req.body.ad !== undefined) templateData.ad = req.body.ad;

    await db
      .update(projects)
      .set({ namingTemplate: JSON.stringify(templateData) })
      .where(eq(projects.id, projectId));

    return res.json({ status: "updated" });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

export default router;
